### ISSUE TITLE
Sampling from a categorical distribution without replacement

### ISSUE THREAD

**Issue by danijar** [tensorflow/tensorflow#9260]
Both `tf.multinomial()` and `tf.contrib.distributions.Categorical.sample()` allow to sample from a multinomial distribution. However, they only allow sampling with replacement.

In constrast, Numpy's `numpy.random.choice()` has a `replace` parameter that allows sampling without replacement. Would it be possible to add a similar functionality to TensorFlow?

One use case is sampling examples from the dataset proportional to the model's last loss on them. When an example generates a very large loss, the next batch will mainly consist of that example. Using sampling without replacement, we can avoid this problem.

I see that sampling with replacement can be parallelized and implemented in a vectorized way, but I don't think sampling speed is a bottleneck in most people's programs.

**Comment by drpngx** [comment]
Sampling without replacement requires book-keeping, and, as you pointed out, in parallel requires some synchronization. I think we would implement per-process semantics, but that might still be slow.
@aselle what do you think?

**Comment by danijar** [comment]
@drpngx I think per-process semantics covers a reasonable fraction of the use cases.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by danijar** [comment]
There is now an easy way to do this:

```python
sample = tf.random_shuffle(tf.range(100))[:10]
sess = tf.Session()
sess.run(sample)  # [10, 83, 99, 88, 56, 77, 39, 48, 32, 41]
sess.run(sample)  # [53, 77, 39, 64, 82,  9, 36, 87, 27, 68]
```

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by mjwillson** [comment]
One relatively GPU-friendly way to do this is with a variant on the Gumbel-max trick (see https://timvieira.github.io/blog/post/2014/07/31/gumbel-max-trick/)

In the standard version (for sampling with replacement), you add i.i.d. Gumbel(0, 1) noise to the logits, then take the argmax index. If instead you want to sample K indices *without* replacement, you can just take the indices of the top K values (e.g. with tf.nn.top_k) instead of the single argmax index.

**Comment by qixiang109** [comment]
@mjwillson haha, thank you, this works! I wrote the codes like:

def tf_no_replacement_sample_using_gambel_max_trick(logits, K):
    z = -tf.log(-tf.log(tf.random_uniform(tf.shape(logits),0,1)))
    _, indices = tf.nn.top_k(logits + z,K)
    return indices

**Comment by qixiang109** [comment]
> Is this equivalent to sampling without replacement or an approximation?

totally equivalent

**Comment by slerman12** [comment]
That’s fantastic. I can sort of intuit why that is, but you wouldn’t happen to have a source for a derivation would you?

**Comment by qixiang109** [comment]
> That’s fantastic. I can sort of intuit why that is, but you wouldn’t happen to have a source for a derivation would you?

well, I can't answer. first of all, what's the derivation of a random sample?

**Comment by slerman12** [comment]
Well, normal sampling without replacing would just be sampling, removing the element, sampling again, and so on. The distribution of that being equivalent to this is non-obvious to me, but I understand why this would at least approximate that.

**Comment by qixiang109** [comment]
haven't prove it but I think the gumbel-max sampling is totally equivalent to a single multinomial sampling,  and the gumbel-top-k-max sampling is also equivalent to a multinomial sampling without replacement. hope someone can give the proof.

**Comment by brianwa84** [comment]
The connection is explained here https://timvieira.github.io/blog/post/2014/08/01/gumbel-max-trick-and-weighted-reservoir-sampling/ and cited from Efraimidis and Spirakis (2005)

**Comment by lorenzoriano** [comment]
I actually tried the code with a small array, and the results are biased:
```
def sample_without_replacement(logits, K):
    """
    Courtesy of https://github.com/tensorflow/tensorflow/issues/9260#issuecomment-437875125
    """
    z = -tf.log(-tf.log(tf.random_uniform(tf.shape(logits),0,1)))
    _, indices = tf.nn.top_k(logits + z, K)
    return indices

number_of_flips = 3
data = tf.constant(list(range(10)), dtype=tf.int32)
with tf.Session() as sess:
    data = tf.cast(data, dtype=tf.float32)
    taken = sample_without_replacement(data, number_of_flips)
    taken = tf_print(taken, [taken], "Taken: ")
    for i in range(10):
        sess.run(taken)
```

The output is:

> Taken:  [9 8 7]
> Taken:  [9 7 6]
> Taken:  [7 9 6]
> Taken:  [7 9 6]
> Taken:  [7 9 8]
> Taken:  [8 9 7]
> Taken:  [9 8 7]
> Taken:  [9 7 8]
> Taken:  [7 9 8]
> Taken:  [9 8 7]

Anything that I'm doing wrong?

**Comment by brianwa84** [comment]
Your logits are 0,1,2,3,4,...
Which says the most likely are those with the highest indices, which is
consistent with the samples. Try logits=tf.zeros([10]) for uniform, or
[0]*5+[1]*5 for something interesting.

**Comment by navarroj** [comment]
For those looking for the simplest case where all elements have the same probably to appear in the sample, i.e. uniform distribution, the formula is just:

```
def random_choice(domain, sample, batch_shape=()):
  """Generate random samples without replacement.
  
  This would be roughly equivalent to:
  
      numpy.random.choice(domain, sample, replace=False)
      
  but also supports generating batches of samples.
  """
  p = tf.random_uniform(batch_shape + (domain,), 0, 1)
  _, indices = tf.nn.top_k(p, sample)
  return indices
```

**Comment by dynamicwebpaige** [comment]
This would be a great feature add, but is outside of the scope of TensorFlow Core and would fall under the jurisdiction of TensorFlow Probability. @jvdillon, could you please transfer the issue?

**Comment by qixiang109** [comment]
> data

logits = log(p), not p .

**Comment by qixiang109** [comment]
> I actually tried the code with a small array, and the results are biased:
> 
> ```
> def sample_without_replacement(logits, K):
>     """
>     Courtesy of https://github.com/tensorflow/tensorflow/issues/9260#issuecomment-437875125
>     """
>     z = -tf.log(-tf.log(tf.random_uniform(tf.shape(logits),0,1)))
>     _, indices = tf.nn.top_k(logits + z, K)
>     return indices
> 
> number_of_flips = 3
> data = tf.constant(list(range(10)), dtype=tf.int32)
> with tf.Session() as sess:
>     data = tf.cast(data, dtype=tf.float32)
>     taken = sample_without_replacement(data, number_of_flips)
>     taken = tf_print(taken, [taken], "Taken: ")
>     for i in range(10):
>         sess.run(taken)
> ```
> 
> The output is:
> 
> > Taken:  [9 8 7]
> > Taken:  [9 7 6]
> > Taken:  [7 9 6]
> > Taken:  [7 9 6]
> > Taken:  [7 9 8]
> > Taken:  [8 9 7]
> > Taken:  [9 8 7]
> > Taken:  [9 7 8]
> > Taken:  [7 9 8]
> > Taken:  [9 8 7]
> 
> Anything that I'm doing wrong?

logits = log(P), not P

**Comment by Randl** [comment]
Is there an open issue in TFP for that?

**Comment by brianwa84** [comment]
I'm not aware of any issue open for this in TFP. Feel free to open one and
send a PR. This could perhaps sit alongside tfp.math.random_rademacher, or
maybe we would want to make a new 'random' pkg in TFP. Team would need to
discuss.

**Comment by nash911** [comment]
> @mjwillson haha, thank you, this works! I wrote the codes like:
> 
> z = -tf.log(-tf.log(tf.random_uniform(tf.shape(logits),0,1)))
> _, indices = tf.nn.top_k(logits + z,K)
> return indices

Interesting approach, but then how do you calculate gradients w.r.t indices?

**Comment by brianwa84** [comment]
Integer-valued tensors don't have a gradient. If you wanted gradients, you'd either relax the categorical (see papers on "concrete") or use a REINFORCE (high variance) gradient. There is also the GO gradient, which can work when the categorical sample is a stochastic leaf.
